### PR TITLE
oracle: make JSONPath compatible with C# implementation

### LINKED ISSUE #1903: Oracle JSONPath filter result differs from result of testnet oracles
neo-go version 0.94.1

I used JSONPath filter `$..item[?(@.id==1)].price` in oracle call of the JSON object stored in NeoFS:

```json
{ "store": {
    "item": [
      { "id": 1,
        "category": "clothing",
        "title": "Aloha Shirt",
        "price": 700
      },
      { "id": 2,
        "category": "clothing",
        "title": "Pearl Tee",
        "price": 400
      },
      { "id": 3,
        "category": "headgear",
        "title": "Fugu Bell Hat",
        "price": 1700
      },
      { "id": 4,
        "category": "headgear",
        "title": "Splash Goggles",
        "price": 2800
      },
      { "id": 5,
        "category": "shoes",
        "title": "Cherry Kicks",
        "price": 2400
      },
      { "id": 6,
        "category": "shoes",
        "title": "Red Hi-Tops",
        "price": 1800
      }
    ]
  }
}
```

In neofs-dev-env environment, single neo-go oracle node returned `[[700]]` string. But oracle nodes in N3 testnet returned `[700]` string.

### PR DESCRIPTION
**Pull request by fyrchik**
Port  neo-project/neo#2446 (WIP).
Fix #1903 .

C# node uses simplified implementation which is easy to port.

Waiting for neo PR to settle.

### CODE REVIEW

File: pkg/services/oracle/filter.go [hunk 1]
```diff
@@ -3,13 +3,445 @@ package oracle
...
 	"github.com/nspcc-dev/neo-go/pkg/core/state"
 	"github.com/nspcc-dev/neo-go/pkg/core/transaction"
 )
 
+type (
+	// pathTokenType represents single JSONPath token.
+	pathTokenType byte
+
+	// pathParser combines JSONPath and a position to start parsing from.
+	pathParser struct {
```
Comment (roman-khimov) [inline pkg/services/oracle/filter.go:20]: Can we move JsonPath implementation to a separate package?

Comment (fyrchik) [reply pkg/services/oracle/filter.go]: Moved.

**Comment by roman-khimov** [comment]
I guess it shouldn't be draft any longer as RC2 is already out.

File: pkg/services/oracle/jsonpath/jsonpath.go [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+	// pathParser combines JSONPath and a position to start parsing from.
+	pathParser struct {
+		s     string
+		i     int
+		depth int
+	}
+)
+
+const (
+	pathRoot pathTokenType = iota
```
Comment (roman-khimov) [inline pkg/services/oracle/jsonpath/jsonpath.go:23]: You can start with `pathInvalid` here.

File: pkg/services/oracle/jsonpath/jsonpath.go [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+		}
+	}
+
+	if objs == nil {
+		objs = []interface{}{}
+	}
+	return objs, true
+}
+
+func (p *pathParser) nextToken() (pathTokenType, string, bool) {
```
Comment (roman-khimov) [inline pkg/services/oracle/jsonpath/jsonpath.go:85]: And then return `pathTokenType, string` here saving on `bool` and some error checks.

File: pkg/services/oracle/jsonpath/jsonpath.go [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+	case pathIdentifier:
+		return p.descentIdent(objs, value)
+	default:
+		return nil, false
+	}
+}
+
+// descent descends 1 level down.
+// It flattens arrays and returns map values for maps.
+func (p *pathParser) descent(objs []interface{}) ([]interface{}, bool) {
```
Comment (roman-khimov) [inline pkg/services/oracle/jsonpath/jsonpath.go:197]: Functions are more like verbs so `descend` can be more appropriate.

File: pkg/services/oracle/jsonpath/jsonpath_test.go [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+
+	t.Run("$[2:1], invalid slice", func(t *testing.T) {
+		_, ok := unmarshalGet(t, "[1,2,3]", "$[2:1]")
+		require.False(t, ok)
+	})
+}
+
+func TestDescentIdent(t *testing.T) {
+	js := `{
+		"store": {
```
Comment (roman-khimov) [inline pkg/services/oracle/jsonpath/jsonpath_test.go:92]: Something's wrong with indentation here.

File: pkg/services/oracle/jsonpath/jsonpath.go [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+		subEnd := end
+		if subEnd <= 0 {
+			subEnd += len(arr)
+		}
+
+		if subEnd > len(arr) {
+			subEnd = len(arr)
+		}
+
+		count := subEnd - subStart
```
Comment (roman-khimov) [inline pkg/services/oracle/jsonpath/jsonpath.go:468]: You don't really need this variable.

File: pkg/services/oracle/jsonpath/jsonpath.go [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+		if count < 0 {
+			return nil, false
```
Comment (roman-khimov) [inline pkg/services/oracle/jsonpath/jsonpath.go:470]: And I think C# returns `true` in this case, https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.take?view=net-5.0

> If count is less than or equal to zero, source is not enumerated and an empty IEnumerable<T> is returned.

**Review by roman-khimov — COMMENTED** [review]
Parser could certainly be improved, it has some code duplication at the moment, but it should work, so it's not worth wasting much time on it.

File: pkg/services/oracle/jsonpath/jsonpath.go (hunk 7 above)
Comment (fyrchik) [reply pkg/services/oracle/jsonpath/jsonpath.go:470]: Yeah, tested this again, fiexd.

File: pkg/services/oracle/jsonpath/jsonpath.go [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+		return p.descentRange(objs, start, int(index))
+	case pathRightBracket:
+		return p.descentRange(objs, start, 0)
+	default:
+		return nil, false
+	}
+}
+
+// descentRange is similar to descend but skips maps and returns sub-slices for arrays.
+func (p *pathParser) descentRange(objs []interface{}, start, end int) ([]interface{}, bool) {
```
Comment (roman-khimov) [inline pkg/services/oracle/jsonpath/jsonpath.go:411]: This still is `descent` for some reason.

**Review by AnnaShaleva — APPROVED** [review]

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**